### PR TITLE
refactor: load dynamic lib in hybridse

### PR DESCRIPTION
**Pull request by dl239**
java sdk cannot extract function symbols from dynamic lib, so we should  load so in hybridse engine

### CODE REVIEW

File: hybridse/src/udf/dynamic_lib_manager.h [hunk 1]
```diff
@@ -0,0 +1,123 @@
...
+
+namespace hybridse {
+namespace udf {
+
+struct DynamicLibHandle {
+    explicit DynamicLibHandle(void* ptr) {
+        handle = ptr;
+        ref_cnt = 1;
+    }
+    void* handle = nullptr;
```
Comment (aceforeverd) [inline hybridse/src/udf/dynamic_lib_manager.h:36]: suffix `_`

File: hybridse/src/udf/dynamic_lib_manager.h [hunk 2]
```diff
@@ -0,0 +1,123 @@
...
+ public:
+    DynamicLibManager() = default;
+
+    base::Status ExtractFunction(const std::string& name, bool is_aggregate, const std::string& file, std::vector<void*>* funs) {
+        if (funs == nullptr) {
+            return {common::kExternalUDFError, "funs is nullptr"};
+        }
+        std::shared_ptr<DynamicLibHandle> so_handle;
+        {
+            std::lock_guard<std::mutex> lock(mu_);
```
Comment (aceforeverd) [inline hybridse/src/udf/dynamic_lib_manager.h:49]: should possible use read_write lock ?  like https://github.com/4paradigm/OpenMLDB/blob/e2b0b51bbcb7381c45723c96112ada171badaf22/src/statistics/query_response_time/deploy_query_response_time.h#L102-L103

File: hybridse/src/udf/dynamic_lib_manager.h [hunk 3]
```diff
@@ -0,0 +1,123 @@
...
+            if (dlclose(so_handle->handle) != 0) {
+                return {common::kExternalUDFError, "dlclose run error. file is " + file};
+            }
+        }
+        return {};
+    }
+
+ private:
+    std::mutex mu_;
+    std::map<std::string, std::shared_ptr<DynamicLibHandle>> handle_map_;
```
Comment (aceforeverd) [inline hybridse/src/udf/dynamic_lib_manager.h:115]: cpplint

File: hybridse/src/udf/udf_library.cc [hunk 4]
```diff
@@ -197,18 +197,41 @@ UdafRegistryHelper UdfLibrary::RegisterUdaf(const std::string& name) {
...
     if (HasFunction(canon_name)) {
         return {kCodegenError, name + " has exist"};
     }
+    std::vector<void*> funs;
+    if (file.empty()) {
+        // use trivial_fun for compile only
+        funs.emplace_back(reinterpret_cast<void*>(udf::v1::trivial_fun));
+    }  else {
+        auto status = lib_manager_.ExtractFunction(canon_name, is_aggregate, file, &funs);
+        if (!status.isOK()) {
```
Comment (aceforeverd) [inline hybridse/src/udf/udf_library.cc]: use marco `CHECK_STATUS` to preserve line info in trace

File: hybridse/src/udf/udf_library.cc [hunk 5]
```diff
@@ -197,18 +197,41 @@ UdafRegistryHelper UdfLibrary::RegisterUdaf(const std::string& name) {
...
+            return status;
+        }
+    }
+    if (funs.empty() || funs[0] == nullptr) {
+        return {kCodegenError, name + " is nullptr"};
```
Comment (aceforeverd) [inline hybridse/src/udf/udf_library.cc]: same here

File: hybridse/src/udf/dynamic_lib_manager.h [hunk 6]
```diff
@@ -0,0 +1,123 @@
...
+        return base::Status::OK();
+    }
+
+    base::Status RemoveHandler(const std::string& file) {
+        std::shared_ptr<DynamicLibHandle> so_handle;
+        {
+            std::lock_guard<std::mutex> lock(mu_);
+            auto iter = handle_map_.find(file);
+            if (iter != handle_map_.end()) {
+                iter->second->ref_cnt--;
```
Comment (aceforeverd) [inline hybridse/src/udf/dynamic_lib_manager.h:99]: not thread safe. Does it matter ?

Comment (dl239) [reply hybridse/src/udf/dynamic_lib_manager.h:99]: decrease is already in lock scope

File: hybridse/src/udf/dynamic_lib_manager.h (hunk 1 above)
Comment (dl239) [reply hybridse/src/udf/dynamic_lib_manager.h:36]: struct members need not end with '_'. refer [Variable_Names](https://google.github.io/styleguide/cppguide.html#Variable_Names)

File: hybridse/src/udf/dynamic_lib_manager.h (hunk 2 above)
Comment (dl239) [reply hybridse/src/udf/dynamic_lib_manager.h:49]: only used in create/drop function. it's not frequently to be called, so it's no matter what's mutex used

**Review by tobegit3hub — APPROVED** [review]
LGTM

**Review by aceforeverd — APPROVED** [review]

**Pull request merged**